extract_embedded_service/__init__.py

```diff
@@ -0,0 +1,148 @@
+        if not clamscan_path:
+            raise ServiceConfigError("You must specify a valid path for clamscan.")
+        if not os.path.isfile(clamscan_path):
+            raise ServiceConfigError("clamscan path does not exist.")
+        if not os.access(clamscan_path, os.X_OK):
+            raise ServiceConfigError("clamscan is not executable.")
+        if not 'clamscan' in clamscan_path.lower():
+            raise ServiceConfigError("Executable does not appear to be clamscan.")
+
+    @staticmethod
+    def get_config(existing_config):
+        # Generate default config from form and initial values.
+        config = {}
+        fields = forms.ExtractEmbeddedConfigForm().fields
+        for name, field in fields.iteritems():
+            config[name] = field.initial
+
+        # If there is a config in the database, use values from that.
+        if existing_config:
+            for key, value in existing_config.iteritems():
+                config[key] = value
+        return config
+
+    @staticmethod
+    def get_config_details(config):
+        return {'clamscan_path': config['clamscan_path']}
+
+    @classmethod
+    def generate_config_form(self, config):
+        html = render_to_string('services_config_form.html',
+                                {'name': self.name,
+                                 'form': forms.ExtractEmbeddedConfigForm(initial=config),
+                                 'config_error': None})
+        form = forms.ExtractEmbeddedConfigForm
+        return form, html
+
+    @staticmethod
+    def valid_for(obj):
+        if not obj.filedata:
+            return False
+        #work for all types
+        return True
+        
+    def run(self, obj, config):
+        obj.filedata.seek(0)
+        data8 = obj.filedata.read(8)
+        obj.filedata.seek(0)
+        user = self.current_task.user
+        self.config = config
+        self.obj = obj
+        self._debug("ExtractEmbedded started")
+        clamscan_path = self.config.get("clamscan_path", "/usr/bin/clamscan")
+        # The _write_to_file() context manager will delete this file at the
+        # end of the "with" block.
+        with self._write_to_file() as tmp_file:
+            (working_dir, filename) = os.path.split(tmp_file)
+            new_env = dict(os.environ)  # Copy current environment
+            args = []
+            obj.filedata.seek(0)
+            #make temp dir for extract embedded file
+            dtmp = tempfile.mkdtemp()
+            #create empty file for no check sig on file
+            emptyrule_path = tempfile.gettempdir() + '/emptyrule.yar'
+            if not os.path.isfile(emptyrule_path):
+                open(emptyrule_path, 'a').close()
+            #TODO add filename filter for avoid escape shell
+            args = [clamscan_path, '--quiet', '--leave-temps', '--tempdir=' + dtmp, '-d', emptyrule_path, filename]
+            if not user.has_access_to(RawDataACL.WRITE):
```

Comment:
Does this service write anything to RawData? If not, does this prevent the user from using the service if they don't have RawData Write permissions even though it doesn't interact with RawData?